bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/BigtableRegionLocator.java

```diff
@@ -61,11 +62,29 @@ public HRegionLocation getRegionLocation(byte[] row) throws IOException {
   /** {@inheritDoc} */
   @Override
   public HRegionLocation getRegionLocation(byte[] row, boolean reload) throws IOException {
-    for (HRegionLocation region : getRegions(reload)) {
-      if (region.getRegionInfo().containsRow(row)) {
-        return region;
+    List<HRegionLocation> regions = getRegions(reload);
+    return findRegion(regions, row);
+  }
+
+  private HRegionLocation findRegion(List<HRegionLocation> regions, byte[] row) throws IOException {
+    int low = 0;
+    int high = regions.size() - 1;
+
+    while (low <= high) {
+      int mid = (low + high) >>> 1;
+      HRegionLocation regionLocation = regions.get(mid);
+      HRegionInfo regionInfo = regionLocation.getRegionInfo();
+
+      // This isn't the last region (endKey != "") and row key is fall outside of it
```

Comment:
// This isn't the last region (endKey != "") and row key is greater than the current bound